
#BiocManager::install("clusterProfiler")
#BiocManager::install("ReactomePA")

# https://www.one-tab.com/page/JUc6Sh1MTFuqRe6uyyi74A

#---------------------------------------------------------------------------------------------------------------------------------------

##* packages needed
library(DiffBind)
library(ChIPseeker)
# library(ReactomePA)
library(clusterProfiler)
library(biomaRt)

library(org.Hs.eg.db)  
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene




#---------------------------------------------------------------------------------------------------------------------------------------

##* path
RES="/home/psurana/projects/atac/FTO_PCDH/"

##* load obj
load(file = paste0(RES, "data/diffBind_overlaps_FTO_PCDH.RData"))
load(file = paste0(RES, "data/diffBind_peaks_FTO_PCDH.RData"))
load(file = paste0(RES, "data/diffBind_FTO_PCDH.RData"))


#---------------------------------------------------------------------------------------------------------------------------------------

##* viewing all samples
dba.show(dbObj.peaks)

# SAMPLES
# 1 2 3 4 5 6
# FTO1 FTO2 FTO3 PCDH1 PCDH2 PCDH3

##* get consensus peaks
db_consensus <- dba(dbObj.peaks, mask=dbObj$masks$Consensus, minOverlap=0.66)

# results.peaks object was the final one containing consensus peaks and differential binding results
peaks_consensus <- dba.peakset(db_consensus, bRetrieve = T)


# extracting FTO peaks
peaks.FTO_rep1 = db_consensus[dbObj.peaks$called[,1]==1] # peaks called in rep 1
peaks.FTO_rep2 = peaks.consensus[db_consensus$called[,2]==1] # peaks called in rep 2
peaks.FTO_rep3 = peaks.consensus[db_consensus$called[,3]==1] # peaks called in rep 3


#---------------------------------------------------------------------------------------------------------------------------------------


##* annotate using CHIPSeeker

PCDH = results.peaks[results.peaks$Fold>0 & results.peaks$FDR<0.05 & results.peaks$'p-value'<0.01]
FTO = results.peaks[results.peaks$Fold<0 & results.peaks$FDR<0.05 & results.peaks$'p-value'<0.01]

##* annotate peaks - PCDH
PCDH.peaksAnno <- annotatePeak(PCDH, tssRegion=c(-1500, 1500), TxDb=txdb, annoDb="org.Hs.eg.db")
# Annotated peaks generated by ChIPseeker
# 1336/1336  peaks were annotated
# Genomic Annotation Summary:
#              Feature  Frequency
# 9           Promoter 23.5778443
# 4             5' UTR  0.2994012
# 3             3' UTR  2.4700599
# 1           1st Exon  0.9730539
# 7         Other Exon  2.8443114
# 2         1st Intron 15.4191617
# 8       Other Intron 23.9520958
# 6 Downstream (<=300)  0.0748503
# 5  Distal Intergenic 30.3892216

write.table(PCDH.peaksAnno, file = paste0(RES, "data/annotated_peaks/Differential_PCDH_from_overlapped.csv"), sep="\t", quote=F, row.names=F)


##* annotate peaks - FTO
FTO.peaksAnno=annotatePeak(FTO, tssRegion=c(-1500, 1500), TxDb=txdb, annoDb="org.Hs.eg.db")
# Annotated peaks generated by ChIPseeker
# 101/101  peaks were annotated
# Genomic Annotation Summary:
#             Feature Frequency
# 7          Promoter  6.930693
# 3            3' UTR  1.980198
# 1          1st Exon  2.970297
# 5        Other Exon  2.970297
# 2        1st Intron 18.811881
# 6      Other Intron 30.693069
# 4 Distal Intergenic 35.643564

write.table(FTO.peaksAnno, file = paste0(RES, "data/annotated_peaks/Differential_FTO_from_overlapped.csv"), sep="\t", quote=F, row.names=F)

##* annotate peaks overall differential list
all.peaksAnno=annotatePeak(results.peaks, tssRegion=c(-1500, 1500), TxDb=txdb, annoDb="org.Hs.eg.db")
100083/100224  peaks were annotated
# Genomic Annotation Summary:
#              Feature  Frequency
# 9           Promoter 25.0242299
# 4             5' UTR  0.2108250
# 3             3' UTR  2.1412228
# 1           1st Exon  0.9991707
# 7         Other Exon  2.6567949
# 2         1st Intron 15.0535056
# 8       Other Intron 26.5899304
# 6 Downstream (<=300)  0.1598673
# 5  Distal Intergenic 27.1644535

write.table(all.peaksAnno, file = paste0(RES, "data/annotated_peaks/Differential_all_from_overlapped.csv"), sep="\t", quote=F, row.names=F)


##* annotations list
list.annotations <- list(FTO=FTO.peaksAnno, PCDH=PCDH.peaksAnno)


#---------------------------------------------------------------------------------------------------------------------------------------


##* plots - annotation distribution
jpeg(file = paste0(RES, "plots/FTO_peaksAnno_distribution.jpeg"))
plotAnnoBar(FTO.peaksAnno, title="FTO differential peaks Feature Distribution")
dev.off()


jpeg(file = paste0(RES, "plots/PCDH_peaksAnno_distribution.jpeg"))
plotAnnoBar(PCDH.peaksAnno, title="PCDH differential peaks Feature Distribution")
dev.off()


jpeg(file = paste0(RES, "plots/All_peaksAnno_distribution.jpeg"))
plotAnnoBar(all.peaksAnno, title="All differential peaks Feature Distribution")
dev.off()


#---------------------------------------------------------------------------------------------------------------------------------------

##* plots - distance to TSS

jpeg(file = paste0(RES, "plots/FTO_TSS_distribution.jpeg"))
plotDistToTSS(FTO.peaksAnno, title="FTO - Distribution of transcription factor-binding loci relative to TSS")
dev.off()


jpeg(file = paste0(RES, "plots/PCDH_TSS_distribution.jpeg"))
plotDistToTSS(PCDH.peaksAnno, title="PCDH - Distribution of transcription factor-binding loci relative to TSS")
dev.off()



jpeg(file = paste0(RES, "plots/PCDH_TSS_distribution.jpeg"))
plotDistToTSS(PCDH.peaksAnno, title="PCDH - Distribution of transcription factor-binding loci relative to TSS")
dev.off()

#---------------------------------------------------------------------------------------------------------------------------------------

##* list plots
jpeg(file = paste0(RES, "plots/PCDH_FTO_TSS_distribution.jpeg"))
plotDistToTSS(list.annotations, title="PCDH and FTO - Distribution of transcription factor-binding loci relative to TSS")
dev.off()


jpeg(file = paste0(RES, "plots/PCDH_FTO_peaksAnno_distribution.jpeg"))
plotAnnoBar(list.annotations, title="PCDH and FTO - differential peaks Feature Distribution")
dev.off()

#---------------------------------------------------------------------------------------------------------------------------------------

genes = lapply(list.annotations, function(i) as.data.frame(i)$geneId)
names(genes) = sub("_", "\n", names(genes))

save(genes, file = paste0(RES, "data/genes_list.RData"))


#---------------------------------------------------------------------------------------------------------------------------------------



# peak annotate for unique peaks using chippeak anno

pcdh.uni <- read.delim("/Users/pallavi/Documents/cl/projects/atac/FTO_PCDH/data/peaks_diffbind/peaks_PCDH_unique.bed", sep="\t")
pcdh.uni = pcdh.uni %>% na.omit() %>% unique()
pcdh.uni = unique(pcdh.uni)

bed2grange = function(i){
  GRanges(seqnames = i$chr,ranges = IRanges(start = i$start,
                                            end = i$end,
                                            names = rownames(i)))
}

pcdh.uni.bed = bed2grange(i=pcdh.uni)

pcdh.uni.peaksAnno=annotatePeak(pcdh.uni.bed, tssRegion=c(-1500, 1500), TxDb=txdb, annoDb="org.Hs.eg.db")
# Annotated peaks generated by ChIPseeker
# 11378/11388  peaks were annotated
# Genomic Annotation Summary:
#   Feature  Frequency
# 9           Promoter 23.1675163
# 4             5' UTR  0.1933556
# 3             3' UTR  2.2851116
# 1           1st Exon  1.3007558
# 7         Other Exon  2.7333451
# 2         1st Intron 15.1256811
# 8       Other Intron 27.7025839
# 6 Downstream (<=300)  0.1757778
# 5  Distal Intergenic 27.3158727

write.table(pcdh.uni.peaksAnno, file = paste0(RES, "data/annotated_peaks/Unique_PCDH_annotated.csv"), sep="\t", quote=F, row.names=F)

##* Similarly for fto gives-

# Annotated peaks generated by ChIPseeker
# 11485/11497  peaks were annotated
# Genomic Annotation Summary:
#   Feature  Frequency
# 9           Promoter 22.3247714
# 4             5' UTR  0.2002612
# 3             3' UTR  1.8894210
# 1           1st Exon  1.2973444
# 7         Other Exon  2.5163256
# 2         1st Intron 15.2285590
# 8       Other Intron 27.7405311
# 6 Downstream (<=300)  0.1218981
# 5  Distal Intergenic 28.6808881

list.annotations <- list(FTO=fto.uni.peaksAnno, PCDH=pcdh.uni.peaksAnno)


jpeg(file = paste0(RES, "plots/Unique_PCDH_FTO_TSS_distribution.jpeg"))
plotDistToTSS(list.annotations, title="Unique -Distribution of transcription factor-binding loci relative to TSS")
dev.off()

jpeg(file = paste0(RES, "plots/Unique_PCDH_FTO_peaksAnno_distribution.jpeg"))
plotAnnoBar(list.annotations, title="Unique - differential peaks Feature Distribution")
dev.off()


#---------------------------------------------------------------------------------------------------------------------------------------
#---------------------------------------------------------------------------------------------------------------------------------------
#---------------------------------------------------------------------------------------------------------------------------------------







